src/python/pants/backend/shell/util_rules/adhoc_process_support.py

```diff
@@ -183,41 +190,68 @@ def _shell_tool_safe_env_name(tool_name: str) -> str:
     return re.sub(r"\W", "_", tool_name)
 
 
-@rule_helper
-async def _shell_command_tools(
-    shell_setup: ShellSetup.EnvironmentAware, tools: tuple[str, ...], rationale: str
-) -> dict[str, str]:
+@dataclass(frozen=True)
+class BashToolsRequest:
+    tools: tuple[str, ...]
+    rationale: str
```

Comment:
This looks very much like `BinaryShimsRequest`..

https://github.com/pantsbuild/pants/blob/b587181465f62b483c95b6f18f4c6cafc90033a9/src/python/pants/core/util_rules/system_binaries.py#L172-L224

Reply:
But I think you're implementation is superior. Mind fixing the binary shims rule, perhaps? :)

Reply:
Well, I'd completely missed `BinaryShims`. That's handy.

What in particular is superior in my implementation? I see, specifically:

* error checking
* using digest operations rather than a bash command to generate the scripts

Is that all?

Reply:
all or more. I didn't compare them in depth, only noticed I had used Processes to create symlinks where you used digest operations.

Reply:
Great, I'll get that working against this branch, and then throw a separate PR over the fence for you to review.

Reply:
See #18184.

Reply:
✅

Reply:
@kaos I've rebased this branch on top of #18184; if you wish to re-review, you can safely ignore the changes in `system_binaries*.py`